src/traces/box/calc.js

```diff
@@ -200,7 +200,13 @@ function calcSelection(cd, trace) {
     if(Array.isArray(trace.selectedpoints)) {
         for(var i = 0; i < cd.length; i++) {
             var pts = cd[i].pts || [];
-            Lib.tagSelected(pts, trace);
+            var ptNumber2cdIndex = {};
+
+            for(var j = 0; j < pts.length; j++) {
+                ptNumber2cdIndex[pts[j].i] = j;
+            }
+
+            Lib.tagSelected(pts, trace, ptNumber2cdIndex);
```

Comment:
Quick fixup. I noticed that `selectedpoints` -> calcdata was off for box and violins while making

![peek 2017-11-17 14-26](https://user-images.githubusercontent.com/6675409/32966063-9ef2064e-cba6-11e7-91ea-a2fbcacbd1c1.gif)

```js
var gen = v => Math.round(Math.random(v) * 10)
var n = 100
var x = new Array(n).fill(0).map(gen)
var y = new Array(n).fill(0).map(gen)
var c = '#984ea3'
var usmo = 0

Plotly.newPlot(gd, [{
  mode: 'markers',
  x: x,
  y: y,
  marker: {color: c},
  unselected: { marker: {opacity: usmo} }
}, {
  type: 'violin',
  name: 'x',
  x: x,
  points: 'all',
  pointpos: 1.5,
  box: {visible: true},
  yaxis: 'y2',
  marker: {color: c},
  unselected: { marker: {opacity: usmo} }
}, {
  type: 'violin',
  name: 'y',
  y: y,
  points: 'all',
  pointpos: 1.5,
  box: {visible: true},
  xaxis: 'x2',
  marker: {color: c},
  unselected: { marker: {opacity: usmo} }
}], {
  xaxis: {
    domain: [0, 0.48]
  },
  yaxis: {
    domain: [0, 0.48]
  },
  xaxis2: {
    domain: [0.52, 1]
  },
  yaxis2: {
    domain: [0.52, 1]
  },
  showlegend: false,
  dragmode: 'lasso',
  hovermode: 'closest',
  width: 500,
  height: 500
})
gd.on('plotly_selected', d => {
  if(d && d.points && d.points.length) {
    var sel = d.points.map(p => p.pointIndex)
    Plotly.restyle(gd, 'selectedpoints', [sel.slice(), sel.slice(), sel.slice()])
  } else {
    Plotly.restyle(gd, 'selectedpoints', null)
  }
})
```